### PR TITLE
Feature integrity check

### PR DESCRIPTION
**Pull request by valydia**
Adding new global directive integrityCheck("filenameColumn","includefolder")
=============================================================

This directive specifies whether or not to perform an integrity check i.e. a check to see if all file under the content folder
are listed in the metadata file. The 1st parameter is the name of the column of the file path and the second optional parameter
is a string that take either "includeFolder" or "excludeFolder" depending on whether the different folders are listed in the metadata files.

### CODE REVIEW

**Comment by adamretter** [comment]
I really still don't think this makes any sense. Integrity Checking is not a function of validating a CSV file. As I suggested before, I think this would be better placed as an implementation specific option, i.e. It really should not be in the grammar.

File: csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/AllErrorsMetaDataValidator.scala [hunk 1]
```diff
@@ -36,20 +36,6 @@ trait AllErrorsMetaDataValidator extends MetaDataValidator {
     v.sequence[MetaDataValidation, Any]
   }
 
-  private def validateRow(row: Row, schema: Schema): MetaDataValidation[Any] = {
```
Comment (adamretter) [inline csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/AllErrorsMetaDataValidator.scala]: Why are these checks removed?

File: csv-validator-core/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+CSV Validator Core
```
Comment (adamretter) [inline csv-validator-core/README.md]: Why is the file added, the behaviour is documented in the CSV Schema project...

File: csv-validator-cmd/README.md [hunk 3]
```diff
@@ -32,10 +32,6 @@ Usage: validate [options] <csv-path> <csv-schema-path>
         Defines the charset encoding used in the CSV file
   -y <value> | --csv-schema-encoding <value>
         Defines the charset encoding used in the CSV Schema file
-  -i <value> | --integrity-check <value>
-        Defines the filename column in the schema for integrity check
-  -d <value> | --include-directory <value>
-        Specifies whether or not the metadata list the directories for integrity check
```
Comment (adamretter) [inline csv-validator-cmd/README.md]: This looks okay but grammar could be improved - 

"Specifies whether or not the metadata list the directories for integrity check" -> "Specifies whether the CSV lists the directories for the integrity check"

File: csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/integrityCheck/WO_95/tech_acq_metadata_v1_WO95Y14B000.csvs [hunk 4]
```diff
@@ -1,5 +1,5 @@
 version 1.0
-@totalColumns 33
+@totalColumns 33 @integrityCheck("file_path","excludeFolder")
```
Comment (adamretter) [inline csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/integrityCheck/WO_95/tech_acq_metadata_v1_WO95Y14B000.csvs]: Why this change? Are Integrity checks enforced by default? If so they should not be, integrity checking is very much specific to TNA and does not make sense for others.

File: csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/integrityCheck/header/schema.csvs [hunk 5]
```diff
@@ -1,4 +1,4 @@
 version 1.0
-@totalColumns  2
+@totalColumns  2 @integrityCheck("Name","includeFolder")
```
Comment (adamretter) [inline csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/integrityCheck/header/schema.csvs]: As above

File: csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/integrityCheck/noheader/schema.csvs [hunk 6]
```diff
@@ -1,4 +1,4 @@
 version 1.0
-@totalColumns  2 @noHeader
+@totalColumns  2 @noHeader @integrityCheck("Name","includeFolder")
```
Comment (adamretter) [inline csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/integrityCheck/noheader/schema.csvs]: As above

File: csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/noContentSchema.csvs [hunk 7]
```diff
@@ -0,0 +1,4 @@
+version 1.0
+@totalColumns  2 @integrityCheck("Name","includeFolder")
```
Comment (adamretter) [inline csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/noContentSchema.csvs]: As above.

File: csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/FailFastMetaDataValidator.scala [hunk 8]
```diff
@@ -60,16 +60,10 @@ trait FailFastMetaDataValidator extends MetaDataValidator {
     validateRows()
   }
 
-  private def validateRow(row: Row, schema: Schema): MetaDataValidation[Any] = {
```
Comment (adamretter) [inline csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/FailFastMetaDataValidator.scala]: What are these checks removed? I can see that you are trying to move them into the base class, but the checks between FailFastMetaDataValiator and AllErrorsMetadataValidator are not the same, which is why they were not in the base class.

**Comment by alexgreenDP** [comment]
You may be right but we need it as its causing difficulties in the pre-ingest phase.

Alex Green
Digital Preservation Services Manager
+44 (0)20 8876 3444 ext 2286
The National Archives, Kew, Richmond, Surrey TW9 4DU
http://www.nationalarchives.gov.uk<http://www.nationalarchives.gov.uk/>

File: csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/api/CsvValidator.scala [hunk 9]
```diff
@@ -74,53 +64,44 @@ trait CsvValidator extends SchemaParser {
...
-   * @return Option[MetaDataValidation] defined only the filename column is defined, and if so return the according metadata
-   */
-  def integrityCheckValidation(csvFile: TextFile, csvSchema: Schema): Option[MetaDataValidation[Any]] = {
-    integrityCheckFilenameColumn.map {filenameColumn =>
-      val columnDedinitionIds = csvSchema.columnDefinitions.map(_.id)
-      if (columnDedinitionIds.exists(_ == filenameColumn)) {
-
-        val filenameColumnIndex = columnDedinitionIds.indexOf(filenameColumn)
-        val filnameColumnRules = csvSchema.columnDefinitions.apply(filenameColumnIndex)
+  def integrityCheckValidation2(csvFile: TextFile, csvSchema: Schema): Option[MetaDataValidation[Any]] = {
```
Comment (adamretter) [inline csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/api/CsvValidator.scala]: Should the integrityCheck not be done in parallel to the CSV validation? Otherwise I fear you will have a long wait of one for the other. Also it seems the integrity check could itself be parallelised.

File: csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/schema/Schema.scala [hunk 10]
```diff
@@ -25,6 +25,8 @@ case class Quoted() extends GlobalDirective("quoted")
 
 case class TotalColumns(numberOfColumns: BigInt) extends GlobalDirective("totalColumns")
 
+case class IntegrityCheck(filepathColumn: String, includeFolder: Boolean) extends GlobalDirective("integrityCheck")
```
Comment (adamretter) [inline csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/schema/Schema.scala]: I don't understand why you want a Global Directive, why are implementation specific command line options not enough?

File: csv-validator-core/src/test/scala/uk/gov/nationalarchives/csv/validator/MetaDataValidatorIntegrityCheckSpec.scala [hunk 11]
```diff
@@ -75,30 +84,43 @@ class MetaDataValidatorIntegrityCheckSpec extends Specification with TestResourc
...
-      validator.validate(TextFile(Path.fromString(headerPath) / "integrityCheckMetaData.csv"), parse(headerPath + "/integrityCheckSchema.csvs",validator), None) must beLike {
+      val validator = buildValidator(substititutionPaths, Some("identifier"))
+      val schema: Schema = parse(noHeaderPath + "/integrityCheckSchema.csvs", validator)
+      validator.validate(TextFile(Path.fromString(noHeaderPath) / "integrityCheckMetaData.csv"), schema, None) must beLike {
         case Success(_) => ok
       }
+
     }
-    
+
```
Comment (adamretter) [inline csv-validator-core/src/test/scala/uk/gov/nationalarchives/csv/validator/MetaDataValidatorIntegrityCheckSpec.scala]: Remove this or mark the test as ignored for Specs

File: csv-validator-core/README.md [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+==================
+
+Core logic for the csv validator.
+
+###Directives
+
+####Global directives
+
+* totalColumns TODO
```
Comment (adamretter) [inline csv-validator-core/README.md]: Please don't commit TODOs in documentation... they never get done...

**Comment by adamretter** [comment]
@alexgreenDP Yup I understand you want some sort of integrity checking and of course I would like to see you guys get what you need. I just fear this approach will cause problems

File: csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/schema/Schema.scala (hunk 10 above)
Comment (valydia) [inline csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/schema/Schema.scala]: Having global directives allows to have more modularity than command line option. For instance, if we want to validate 2 files (metadata and closure), we can enable it for one and not the other using directive within their schemas, this wouldn't be possible (or entail much more changes in the code that runs it) to do so.

File: csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/FailFastMetaDataValidator.scala (hunk 8 above)
Comment (valydia) [inline csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/FailFastMetaDataValidator.scala]: Some of these methods where duplicated in the base class and the AllErrorsMetaDataValidator class. So to avoid duplicate code, i refactored some methods visibility into protected instead of private, and overrode the subclass implementation when needed.

File: csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/schema/Schema.scala (hunk 10 above)
Comment (adamretter) [inline csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/schema/Schema.scala]: @valydia I can't see your calling code so I don't yet really understand Your argument. Your metadata and closure files are different CSV files, and therefore you will have different CSV Schema files. 

I assume your code looks a bit like -

```scala
CSVValidator.create(options).validate(csvs, csv)
```

So you will be creating two validators anyway right? One for the metadata file and one for the closure file? If so, why do you also need the global directive? You could just pass the parameters for your integrity check in the options.

File: csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/FailFastMetaDataValidator.scala (hunk 8 above)
Comment (adamretter) [inline csv-validator-core/src/main/scala/uk/gov/nationalarchives/csv/validator/FailFastMetaDataValidator.scala]: @valydia Ah yes I see, sorry you are correct. I was looking at the wrong context. This bit looks fine.

File: csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/integrityCheck/WO_95/tech_acq_metadata_v1_WO95Y14B000.csvs (hunk 4 above)
Comment (valydia) [inline csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/integrityCheck/WO_95/tech_acq_metadata_v1_WO95Y14B000.csvs]: The integrity check is not enforced by default. If there is no global directive it will behave like before.

Comment (adamretter) [inline csv-validator-cmd/src/test/resources/uk/gov/nationalarchives/csv/validator/cmd/integrityCheck/WO_95/tech_acq_metadata_v1_WO95Y14B000.csvs]: Ah! I see now this is in the `integrityCheck` folder - I thought you were modifying existing tests. All fine.
